File: lib/core/reader/reader_engine_registry.dart
import '../model/book_format.dart';
import 'reader_engine.dart';

class ReaderEngineRegistry {
  final Map<BookFormat, ReaderEngine> _engines;

  ReaderEngineRegistry(List<ReaderEngine> engines)
      : _engines = {for (final e in engines) e.format: e};

  ReaderEngine get(BookFormat format) {
    final engine = _engines[format];
    if (engine == null) {
      throw StateError('No reader engine registered for $format');
    }
    return engine;
  }
}

File: core/reader/src/main/kotlin/com/readler/core/reader/ReaderEngineRegistry.kt
package com.readler.core.reader

import com.readler.core.model.BookFormat

class ReaderEngineRegistry(
    readerEngines: List<ReaderEngine>
) {
    private val enginesByFormat: Map<BookFormat, ReaderEngine> = readerEngines.associateBy { it.format }

    fun get(format: BookFormat): ReaderEngine {
        return requireNotNull(enginesByFormat[format]) { "No reader engine registered for $format" }
    }
}
